File: examples/typescript/02-outbound-lead-workflow-agent/record-fixtures.ts
// Record step (OFF in CI). Authors the canned mock-provider fixtures from the leads set so the example
// runs offline. Re-run only when a prompt or a lead input changes:
//
//   node --experimental-strip-types record-fixtures.ts
//
// There is no real-provider call here — the canned enrichments/scores ARE the leads-set labels, which
// is what makes the run deterministic and the eval a test of the workflow/budget/HITL logic, not the
// model. Note the recorded `usage` token counts: they are the input to the cost model, so lead l3's
// deliberately fat enrich response is what drives it over the per-lead budget cap at the score step.

import { readFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { record } from "../../../tools/mock-llm/index.ts";
import { enrichRequest, scoreRequest, type Lead } from "./src/agent.ts";
import type { Enrichment } from "./src/schema.ts";

interface LeadFixture extends Lead {
  enrich: Enrichment;
  enrichUsage: { input_tokens: number; output_tokens: number };
  score: { fit: number; recommended_action: string };
  scoreUsage: { input_tokens: number; output_tokens: number };
}

const here = path.dirname(fileURLToPath(import.meta.url));
const leads: LeadFixture[] = JSON.parse(readFileSync(path.join(here, "eval", "leads.json"), "utf-8"));
const fixturesDir = path.join(here, "fixtures");

let n = 0;
for (const lead of leads) {
  // enrich fixture
  record(
    enrichRequest(lead),
    { content: JSON.stringify(lead.enrich), stop_reason: "end_turn", usage: lead.enrichUsage },
    fixturesDir,
  );
  n++;
  // score fixture (request includes the enriched object, so it must be recorded with the SAME shape)
  record(
    scoreRequest(lead, lead.enrich),
    { content: JSON.stringify(lead.score), stop_reason: "end_turn", usage: lead.scoreUsage },
    fixturesDir,
  );
  n++;
}
console.log(`recorded ${n} fixtures into ${path.relative(here, fixturesDir)}/`);

File: examples/python/04-research-synthesis-multi-agent/record_fixtures.py
"""Record step (OFF in CI). Authors the canned mock-provider fixtures for example 04 from the golden
set, so the run is offline + deterministic. Re-run only when a prompt or the golden inputs change:

    .venv/bin/python record_fixtures.py

There is no real-provider call here — the canned worker claims, themes, and synthesis ARE the golden
labels. That is what makes the run deterministic and the eval a test of the orchestration + grounding
logic, not the model. The fixtures are keyed by request hash, so the prompt builders in src/agents.py
must produce byte-identical requests at record-time and run-time.
"""
from __future__ import annotations

import json
import os
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
REPO = os.path.abspath(os.path.join(HERE, "..", "..", ".."))
sys.path.insert(0, os.path.join(REPO, "tools", "mock-llm"))
sys.path.insert(0, HERE)

import mock_llm  # noqa: E402
from src import agents  # noqa: E402

FIXTURES = os.path.join(HERE, "fixtures")


def _resp(content_obj, in_tok, out_tok):
    return {
        "content": json.dumps(content_obj, sort_keys=True, separators=(",", ":")),
        "stop_reason": "end_turn",
        "usage": {"input_tokens": in_tok, "output_tokens": out_tok},
    }


# Two golden claims per source — 8 claims total, all grounded by construction (each cites its source).
GOLDEN_CLAIMS = {
    "src-obs": [
        "Emitting one trace event per agent step lets teams reconstruct any run end to end.",
        "Per-step tracing is the biggest lever on mean-time-to-diagnose for agent incidents.",
    ],
    "src-eval": [
        "A regression gate that blocks releases below an eval pass-rate threshold catches prompt drift.",
        "Treating the golden set as a unit test for the agent surfaces defects before they ship.",
    ],
    "src-cost": [
        "Attributing token usage per agent role makes runaway orchestrator loops visible in the bill.",
        "A per-role token budget is the cheapest cost guardrail and the first to pay off.",
    ],
    "src-trust": [
        "A groundedness check rejects any synthesized claim that lacks a source reference.",
        "Requiring a citation per claim stops the agent from asserting facts it cannot cite.",
    ],
}

# One synthesized claim per theme, each citing the source ids it draws on (kept in corpus → grounded).
GOLDEN_THEMES_ORDER = [
    ("Observability", "src-obs"),
    ("Evaluation", "src-eval"),
    ("Cost", "src-cost"),
    ("Trust", "src-trust"),
]


def main():
    golden = json.load(open(os.path.join(HERE, "eval", "golden.json"), encoding="utf-8"))
    sources = golden["sources"]

    n = 0
    # 1) Worker fixtures: one per source. The response is the golden claim set for that source.
    claims = []
    for source in sources:
        req = agents.worker_request(source)
        claim_objs = [{"text": t, "source": source["id"]} for t in GOLDEN_CLAIMS[source["id"]]]
        mock_llm.record(req, _resp({"claims": claim_objs}, 90, 40), FIXTURES)
        n += 1
        for c in claim_objs:
            claims.append({"text": c["text"], "source": source["id"], "worker": source["worker"]})

    # 2) Theme-grouping fixture: groups the flat claim list (built from the worker outputs above) into
    #    one theme per source, covering each claim exactly once.
    by_source_indexes = {}
    for i, c in enumerate(claims):
        by_source_indexes.setdefault(c["source"], []).append(i)
    themes = [{"name": name, "claim_indexes": by_source_indexes[sid]} for name, sid in GOLDEN_THEMES_ORDER]
    mock_llm.record(agents.group_request(claims), _resp({"themes": themes}, 160, 60), FIXTURES)
    n += 1

    # 3) Synthesis fixture: carry every grounded claim through to the answer (8 total), each tagged
    #    with its theme and citing its source id — so the groundedness eval runs over all 8 claims.
    synth_claims = []
    for name, sid in GOLDEN_THEMES_ORDER:
        for idx in by_source_indexes[sid]:
            synth_claims.append({"text": f"{name}: " + claims[idx]["text"], "sources": [sid]})
    answer = (
        "Reliable production agents rest on four load-bearing practices: per-step observability, "
        "evaluation gates, per-role cost attribution, and a groundedness check on every claim."
    )
    mock_llm.record(
        agents.synth_request(themes, claims),
        _resp({"answer": answer, "claims": synth_claims}, 220, 90),
        FIXTURES,
    )
    n += 1

    print(f"recorded {n} fixtures into {os.path.relpath(FIXTURES, HERE)}/")


if __name__ == "__main__":
    main()
